### DISCUSSION TITLE
Upload video via API?

### DISCUSSION THREAD

**Discussion by Cpalen17** [MarshalX/atproto#426] [Questions] [answered]
I am trying to bring over my Twitter bot to Bluesky, the biggest part is the tweets come with a short 10-20 second video of a baseball play. Using the below screen shot is exactly how I found how to do this but doesnt seem to actually work as the mp4 doesn't get uploaded. 

Anyone know about this?

![image](https://github.com/user-attachments/assets/579f9d68-d3c4-470c-8e26-40bf6ad236ab)

**Comment by emilyhunt** [comment]
What happens when you try to upload the video? Do you get an error message? Is your client logged in?

Also, as a test - are you able to send normal posts without video ok, and is it just `send_video` causing the issue?

You've probably already found it but [the example here](https://github.com/MarshalX/atproto/blob/main/examples/send_video.py) should work.

**Reply by moniesoke** [reply]
Hello,
Thanks for the tips. I have the same issue.
I can use client.send_post send normal text posts. 
client.send_video also works but the video says "Video not found". 
Here is one of the post I tried.
https://bsky.app/profile/videoofchina.bsky.social/post/3lb4ph2rh4l2l
I tried asynch version too. same result:
https://bsky.app/profile/videoofchina.bsky.social/post/3lb4r7qh3ng2q

Send_image works fine.

I also tried using blob upload (com.atproto.repo.uploadBlob) and use 'com.atproto.repo.createRecord' to create post with embed but I tried a few things the video doesn't show up in the post, only text shows up.

**Reply by Cpalen17** [reply]
https://bsky.app/profile/catchprobability.bsky.social

On this profile you can see. I actually had a successful upload. Then had it wait 30 seconds and try again and then a different video failed

**Comment by MarshalX** [comment, marked as answer]
You baited me to show to properly use Video Service with async video processing. Which could return useful error message exactly why you video failed to process. Which is missed in the `uploadBlob` method. But I've faced with bunch of server side errors. You can track the progress here: https://github.com/MarshalX/atproto/pull/436

**Reply by moniesoke** [reply]
Thanks for the response! I noticed that it is probably caused by the fact that I haven't verify my email. 
Even before verifying my email, I can use API to send text posts. But I cannot send video posts. 
After verifying my email, the send_video works fine. Thanks!

**Reply by MarshalX** [reply]
Ah, yeah... This is kind of the entry level gate to not blow up video processing service. Now I will recommend to everyone to verify video uploading from the web and only then try to do with API. Thank you!

**Reply by moniesoke** [reply]
> Ah, yeah... This is kind of the entry level gate to not blow up video processing service. Now I will recommend to everyone to verify video uploading from the web and only then try to do with API. Thank you!

Thanks! I have mixed successes with the send_video function. I think it doesn't work well for longer videos even though it is below 60 seconds long.  I cut video to 59 seconds. and send_video still cannot post it properly (only text and embedded black screen) and I got "Video Not Found" on the embedded video. But if I upload it manually through webpage it works fine. So I guess the API still have some issues. When I manually upload, it takes time to process the video like 5 seconds. Using send_video API, it seems instantaneous but the video doesn't show up. 

Is there a guide on how to load the video step by step. like upload blob, process, and embed in a post? So I may discover where the problem is. Thanks!

**Reply by MarshalX** [reply]
Check my shared PR that I have mentioned above. This is different upload flow which in use by official bsky app. It is not working 100% properly. That's why it is not merged. Waiting for the answer from bsky team

**Reply by moniesoke** [reply]
One of the video is about 25MB 59 seconds long. send_video shows a timeout error. I noticed that the video that failed to publish has 60 fps so I  reduced to 30 fps. still doesn't work. Noticed that it has > 2000kbps so reduced to 1000kbps. File size reduced from 25MB to about 8MB. Finally works ok. Not sure whether the bit rate reduction is helping or the file size reduction is helping.

**Reply by MarshalX** [reply]
@moniesoke you can increase timeout. I shown how to do it here https://github.com/MarshalX/atproto/discussions/419#discussioncomment-11283766

**Reply by moniesoke** [reply]
Thanks! Everything works now. I will try the timeout setting. maybe I won't need to compress videos that much.
